buildings/reference_data/other_reference.py

```diff
@@ -191,6 +195,105 @@ def update_imagery_survey_index(kx_api_key, dataset, dbconn):
     return "updated"
 
 
+def update_facilities(kx_api_key, dataset, dbconn):
+    if dataset != "nz_facilities":
+        return "error"
+
+    # get last update of layer date from log
+    from_var = last_update("facilities")
+
+    # current date
+    to_var = current_date()
+
+    cql_filter = ""
+    external_id = "facility_id"
+
+    layer = QgsVectorLayer(
+        URI.format(LDS_LAYER_IDS[dataset], kx_api_key, from_var, to_var, cql_filter)
+    )
+    if not layer.isValid():
+        # something went wrong
+        return "error"
+
+    if layer.featureCount() == 0:
+        return "current"
+
+    for feature in layer.getFeatures():
+        if feature.attribute("__change__") == "DELETE":
+            sql = "DELETE FROM buildings_reference.nz_facilities WHERE facility_id = %s;"
+            dbconn.execute_no_commit(sql, (feature[external_id],))
+
+        elif feature.attribute("__change__") == "INSERT":
+            sql = "SELECT True FROM buildings_reference.nz_facilities WHERE facility_id = %s;"
+            result = dbconn.execute_return(
+                sql,
+                (feature[external_id],),
+            )
+            result = result.fetchone()
+            if result is None:
+                sql = """  
+                    INSERT INTO buildings_reference.nz_facilities (
+                        facility_id,
+                        source_facility_id,
+                        name,
+                        source_name,
+                        use,
+                        use_type,
+                        use_subtype,
+                        estimated_occupancy,
+                        last_modified,
+                        shape)
+                    VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, ST_SetSRID(ST_GeometryFromText(%s), 2193))
+                    """
+                dbconn.execute_no_commit(
+                    sql,
+                    (
+                        feature[external_id],
+                        correct_attribute_format(feature["source_facility_id"]),
+                        correct_attribute_format(feature["name"]),
+                        correct_attribute_format(feature["source_name"]),
+                        correct_attribute_format(feature["use"]),
+                        correct_attribute_format(feature["use_type"]),
+                        correct_attribute_format(feature["use_subtype"]),
+                        correct_attribute_format(feature["estimated_occupancy"]),
+                        feature["last_modified"].toString(Qt.ISODate),
+                        feature.geometry().asWkt(),
+                    ),
+                )
+
+        elif feature.attribute("__change__") == "UPDATE":
```

Comment:
we could add error checking to report when a feature wasnt updated. Is the assumption we don't need this as has already been checked before creating this facilities dataset?

Reply:
Theoretically that shouldn't happen. The data is from the LDS which should be clean, and the table shouldn't get changed manually because it's only reference data. Other reference data works the same way, so I wouldn't worry about adding that error check right now.